libmachine/provision/configure_swarm.go

```diff
@@ -41,70 +43,80 @@ func configureSwarm(p Provisioner, swarmOptions swarm.Options, authOptions auth.
 	advertiseInfo := fmt.Sprintf("%s:%s", ip, dockerPort)
 
 	if swarmOptions.Master {
-		cmd := fmt.Sprintf("manage --tlsverify --tlscacert=%s --tlscert=%s --tlskey=%s -H %s --strategy %s --advertise %s",
-			authOptions.CaCertRemotePath,
-			authOptions.ServerCertRemotePath,
-			authOptions.ServerKeyRemotePath,
-			swarmOptions.Host,
-			swarmOptions.Strategy,
-			advertiseInfo,
-		)
-
-		cmdMaster := strings.Fields(cmd)
-		for _, option := range swarmOptions.ArbitraryFlags {
-			cmdMaster = append(cmdMaster, "--"+option)
+		err = configureSwarmMaster(dockerHost, swarmOptions, authOptions, dockerDir, port, advertiseInfo)
+		if err != nil {
+			return err
 		}
+	}
 
-		//Discovery must be at end of command
-		cmdMaster = append(cmdMaster, swarmOptions.Discovery)
-
-		hostBind := fmt.Sprintf("%s:%s", dockerDir, dockerDir)
-		masterHostConfig := dockerclient.HostConfig{
-			RestartPolicy: dockerclient.RestartPolicy{
-				Name:              "always",
-				MaximumRetryCount: 0,
-			},
-			Binds:        []string{hostBind},
-			PortBindings: map[string][]dockerclient.PortBinding{"3376/tcp": {{"", port}}},
-			NetworkMode:  "host",
-		}
+	return configureSwarmAgent(dockerHost, swarmOptions, advertiseInfo)
+}
 
-		swarmMasterConfig := &dockerclient.ContainerConfig{
-			Image: swarmOptions.Image,
-			Env:   swarmOptions.Env,
-			ExposedPorts: map[string]struct{}{
-				"2375/tcp": {},
-				"3376/tcp": {},
-			},
-			Cmd:        cmdMaster,
-			HostConfig: masterHostConfig,
-		}
+func configureSwarmMaster(dockerHost mcndockerclient.DockerHost, swarmOptions swarm.Options, authOptions auth.Options, dockerDir string, port string, advertiseInfo string) error {
+	cmd := fmt.Sprintf("manage --tlsverify --tlscacert=%s --tlscert=%s --tlskey=%s -H %s --strategy %s --advertise %s",
+		authOptions.CaCertRemotePath,
+		authOptions.ServerCertRemotePath,
+		authOptions.ServerKeyRemotePath,
+		swarmOptions.Host,
+		swarmOptions.Strategy,
+		advertiseInfo,
+	)
+
+	cmdMaster := strings.Fields(cmd)
```

Comment:
Since these are going to become a string slice anyway now, maybe you could instantiate the `cmd` as a `[]string` directly instead of splitting it here?